core/src/data/rasterSource.cpp

```diff
@@ -157,20 +150,29 @@ std::shared_ptr<TileTask> RasterSource::createTask(TileID _tileId, int _subTask)
     return task;
 }
 
-Raster RasterSource::getRaster(const TileTask& _task) {
-    const auto& taskTileID = _task.tileId();
-    TileID id(taskTileID.x, taskTileID.y, taskTileID.z);
+Raster RasterSource::getRaster(const RasterTileTask& _task) {
+    const auto& tileId = _task.tileId();
+    TileID id(tileId.x, tileId.y, tileId.z);
 
-    auto texIt = m_textures->find(id);
-    if (texIt != m_textures->end()) {
-        auto&& texture = texIt->second.lock();
-        return { id,  texture };
+    if (_task.m_texture == m_emptyTexture) {
+        return Raster{id, m_emptyTexture};
     }
 
-    auto& task = static_cast<const RasterTileTask&>(_task);
-    m_textures->emplace(id, task.m_texture);
+    auto entry = m_textures.emplace(id, _task.m_texture);
+    std::shared_ptr<Texture> texture = entry.first->second;
+    LOGD("Cache: add %d/%d/%d - reused: %d", id.x, id.y, id.z, !entry.second);
+
+    // Remove textures that are only held by cache
+    for(auto it = m_textures.begin(), end = m_textures.end(); it != end; ) {
+        if (it->second.use_count() == 1) {
```

Comment:
Ahh, I should have caught this issue before (https://github.com/tangrams/tangram-es/pull/1944/files#diff-10249f514974ffe886f51c3c334bf004L188)!

Reply:
It seems counter intuitive to remove raster textures corresponding to other tileIDs when `getRaster` is called to emplace a raster texture for given tileID (rasterTask).

To me the old code (refer link above) was straight forward in removing all raster textures associated with a tile when it gets deleted.